src/MOI/callbacks/interface.jl

```diff
@@ -0,0 +1,84 @@
+@doc raw"""
+    CallbackData
+
+!!! warn
+    All subtypes from [`CallbackData`](@ref) must be mutable.
+"""
+abstract type CallbackData end
+
+@doc raw"""
+    GenericCallbackData <: CallbackData
+"""
+mutable struct GenericCallbackData <: CallbackData
+    root_model::XpressProblem
+    node_model::Union{XpressProblem, Nothing}
+    data::Any
+
+    function GenericCallbackData(root_model::XpressProblem, data::Any=nothing)
+        return new(root_model, nothing, data)
+    end
+end
+
+@doc raw"""
+    CallbackDataWrapper{CD}
+
+This struct is nothing but a way to inject the julia function provided by the user into the Xpress callback.
+"""
+mutable struct CallbackDataWrapper{CD<:CallbackData}
+    func::Function
+    data::CD
+
+    function CallbackDataWrapper{CD}(root_model::XpressProblem, func::Function, data::Any) where {CD}
+        return new(func, CD(root_model, data))
+    end
+end
+
+function CallbackDataWrapper(root_model::XpressProblem, func::Function, data::Any=nothing)
+    return CallbackDataWrapper{GenericCallbackData}(root_model, func, data)
+end
+
+Base.cconvert(::Type{Ptr{Cvoid}}, x::CallbackDataWrapper) = x
+
+function Base.unsafe_convert(::Type{Ptr{Cvoid}}, x::CallbackDataWrapper)
+    return pointer_from_objref(x)::Ptr{Cvoid}
+end
```

Comment:
If we can do it explicitly without requiring these functions it could be good. as this is hard to debug

Reply:
This was in the code before the rework. I'm not sure how things will be if we change this but I will take a look

Reply:
Please keep as-is. Using `cconvert` and `unsafe_convert` like this is correct. 

Creating the pointer not using these functions is dangerous because Julia could GC the `CallbackDataWrapper` while the solver is inn the callback.